dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/RedshiftClusterUsageMetricsTask.java

```diff
@@ -0,0 +1,196 @@
+import com.amazonaws.services.redshift.model.Cluster;
+import com.amazonaws.services.redshift.model.DescribeClustersRequest;
+import com.google.auto.value.AutoValue;
+import com.google.common.io.ByteSink;
+import com.google.edwmigration.dumper.application.dumper.connector.ZonedInterval;
+import com.google.edwmigration.dumper.application.dumper.handle.Handle;
+import com.google.edwmigration.dumper.application.dumper.task.TaskRunContext;
+import java.io.IOException;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+import javax.annotation.Nonnull;
+import org.apache.commons.lang3.tuple.Pair;
+
+/** Extraction task to get Redshift time series metrics from AWS CloudWatch API. */
+public class RedshiftClusterUsageMetricsTask extends AbstractAwsApiTask {
+
+  protected static enum MetricName {
+    CPUUtilization,
+    PercentageDiskSpaceUsed
+  }
+
+  protected static enum MetricType {
+    Maximum,
+    Average
+  }
+
+  @AutoValue
+  protected abstract static class MetricConfig {
+
+    public abstract MetricName name();
+
+    public abstract MetricType type();
+
+    public static MetricConfig create(MetricName name, MetricType type) {
+      return new AutoValue_RedshiftClusterUsageMetricsTask_MetricConfig(name, type);
+    }
+  }
+
+  private static final String REDSHIFT_NAMESPACE = "AWS/Redshift";
+  private static final DateTimeFormatter DATE_FORMAT =
+      DateTimeFormatter.ISO_OFFSET_DATE_TIME.withZone(ZoneOffset.UTC);
+
+  private final ZonedDateTime currentTime;
+  private final ZonedInterval interval;
+  private final List<MetricConfig> metrics;
+
+  public RedshiftClusterUsageMetricsTask(
+      AWSCredentialsProvider credentialsProvider,
+      ZonedDateTime currentTime,
+      ZonedInterval interval,
+      String zipEntryName,
+      Class<? extends Enum<?>> headerEnum,
+      List<MetricConfig> metrics) {
+    super(credentialsProvider, zipEntryName, headerEnum);
+    this.interval = interval;
+    this.metrics = metrics;
+    this.currentTime = currentTime;
+  }
+
+  @Override
+  protected Void doRun(TaskRunContext context, @Nonnull ByteSink sink, Handle handle)
+      throws IOException {
+
+    List<Cluster> clusters = listClusters();
+    return writeRecordsCsv(
+        sink,
+        clusters.stream()
+            .flatMap(
+                cluster -> {
+                  return zipMetricStreams(
+                          metrics.stream()
+                              .map(
+                                  metricConfig -> {
+                                    return getMetricDataPoints(
+                                        cluster.getClusterIdentifier(), metricConfig);
+                                  })
+                              .collect(toImmutableList()))
+                      .map(
```

Comment:
It looks the two mapping functions can be grouped together without being collected first?

Reply:
I think we cannot merge because first `map` generates data for the single metric, then data is zipped and then second map applies additional column to the result.